### PR TITLE
dyngo: dynamically register listeners only if they're needed

### PR DESCRIPTION
**Pull request by RomainMuller**
### What does this PR do?

With a somewhat significant refactor, and use of Go generics, it is possible to have AppSec listeners self-install if, **and only if** a contrib that may emit events they listen to is loaded.

### Motivation

This avoid logging uncanny messages in debug mode, where we previously claimed to register all available listeners (GraphQL, gRPC, HTTP), despite some of them being nonsensical in the user's application context.

It also allows the compiler to perform better dead code elimination, or even entirely skipping packages corresponding to the unneeded listeners.

### Reviewer's Checklist

- [X] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering this feature have been added and enabled with the va.b.c-dev version tag.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.
- [ ] Add an appropriate team label so this PR gets put in the right place for the release notes.

For Datadog employees:

- [ ] If this PR touches code that handles credentials of any kind, such as Datadog API keys, I've requested a review from `@DataDog/security-design-and-guidance`.
- [X] This PR doesn't touch any of that.

Unsure? Have a question? Request a review!

### CODE REVIEW

File: internal/appsec/dyngo/operation.go [hunk 1]
```diff
@@ -34,50 +33,23 @@ import (
...
-	emitEvent(argsType reflect.Type, op Operation, v interface{})
-
-	emitData(argsType reflect.Type, v any)
-
-	// add the given event listeners to the operation.
-	// add is a private method implemented by the operation struct type so
-	// that no other package can define it.
-	add(...EventListener)
+	// self is an internal method guaranteeing only *operation implements Operation.
+	self() *operation
```
Comment (Julio-Guerra) [inline internal/appsec/dyngo/operation.go:40]: Nice simplification indeed 👏
But while I was reading its use in the code, I was thinking `o.implementation()` or `o.unwrap()` might be more understandable maybe:
```suggestion
	 unwrap() *operation
```

No strong opinion here.

File: internal/appsec/dyngo/operation.go [hunk 2]
```diff
@@ -34,50 +33,23 @@ import (
...
-	// use to emit the finish event.
-	// finish is a private method implemented by the operation struct type so
-	// that no other package can define it.
-	finish(op Operation, results interface{})
+type ArgOf[O Operation] interface {
+	IsArgOf(O)
+}
+type ResultOf[O Operation] interface {
+	IsResultOf(O)
 }
```
Comment (Julio-Guerra) [inline internal/appsec/dyngo/operation.go:48]: Worth some rationals on how this will work :-)

File: internal/appsec/waf.go [hunk 3]
```diff
@@ -115,56 +93,10 @@ func newWAFHandle(rules config.RulesFragment, cfg *config.Config) (*wafHandle, e
...
-	}
-
-	if len(httpAddresses) > 0 {
-		log.Debug("appsec: creating http waf event listener of the rules addresses %v", httpAddresses)
-		listeners = append(listeners, httpsec.NewWAFEventListener(waf.Handle, waf.actions, httpAddresses, cfg.WAFTimeout, &cfg.APISec, l))
-	}
+var wafEventListeners []wafEventListener
 
-	return listeners, nil
+func AddWAFEventListener(fn wafEventListener) {
```
Comment (Hellzy) [inline internal/appsec/waf.go:100]: Nit: short description to keep exported functions documented

File: internal/appsec/listener/grpcsec/grpc.go [hunk 4]
```diff
@@ -32,131 +33,156 @@ const (
...
+var supportedAddresses = listener.AddressSet{
 	GRPCServerRequestMessage:  {},
 	GRPCServerRequestMetadata: {},
 	HTTPClientIPAddr:          {},
 	UserIDAddr:                {},
 }
 
-func SupportedAddressCount() int {
-	return len(supportedAddresses)
+func Install(wafHandle *waf.Handle, actions sharedsec.Actions, cfg *config.Config, lim limiter.Limiter, root dyngo.Operation) {
```
Comment (Hellzy) [inline internal/appsec/listener/grpcsec/grpc.go:43]: Nit: short description to keep exported functions documented

File: internal/appsec/dyngo/operation.go [hunk 5]
```diff
@@ -141,107 +129,103 @@ func NewRootOperation() Operation {
...
 	defer o.mu.RUnlock()
 	if o.disabled {
 		return
 	}
-	o.dataBroadcaster.add(l.ListenedType(), l)
+	addDataListener(&o.dataBroadcaster, l)
 }
 
-func (o *operation) EmitData(data any) {
+func EmitData[T any](op Operation, data T) {
```
Comment (Hellzy) [inline internal/appsec/dyngo/operation.go:226]: Nit: description to keep exported functions documented.
Sorry that one is on me 🙈

File: internal/appsec/dyngo/operation.go [hunk 6]
```diff
@@ -98,29 +80,35 @@ func SwapRootOperation(new Operation) {
...
 
 	disabled bool
 	mu       sync.RWMutex
 }
 
+func (o *operation) Parent() Operation {
+	return o.parent
+}
+
+// This is the one true Operation implementation!
```
Comment (Hellzy) [inline internal/appsec/dyngo/operation.go:95]: All hail 🙏

**Review by Hellzy — COMMENTED** [review]
Cool rework!

 ̶I̶'̶m̶ ̶t̶h̶i̶n̶k̶i̶n̶g̶ ̶w̶e̶ ̶c̶a̶n̶ ̶g̶e̶t̶ ̶r̶i̶d̶ ̶o̶f̶ ̶I̶s̶A̶r̶g̶O̶r̶f̶/̶I̶s̶R̶e̶s̶u̶l̶t̶O̶f̶ ̶i̶m̶p̶l̶e̶m̶s̶ ̶o̶v̶e̶r̶a̶l̶l̶ ̶a̶n̶d̶ ̶j̶u̶s̶t̶ ̶u̶s̶e̶ ̶t̶h̶e̶ ̶i̶n̶t̶e̶r̶f̶a̶c̶e̶s̶ ̶t̶o̶ ̶e̶n̶f̶o̶r̶c̶e̶ ̶t̶y̶p̶e̶ ̶c̶o̶n̶s̶t̶r̶a̶i̶n̶t̶,̶ ̶l̶e̶t̶ ̶m̶e̶ ̶k̶n̶o̶w̶ ̶w̶h̶a̶t̶ ̶y̶o̶u̶ ̶t̶h̶i̶n̶k̶.̶ ̶M̶a̶i̶n̶l̶y̶ ̶i̶t̶ ̶j̶u̶s̶t̶ ̶r̶e̶m̶o̶v̶e̶s̶ ̶t̶h̶e̶ ̶n̶e̶e̶d̶ ̶t̶o̶ ̶i̶m̶p̶l̶e̶m̶e̶n̶t̶ ̶t̶h̶o̶s̶e̶ ̶e̶m̶p̶t̶y̶ ̶f̶u̶n̶c̶t̶i̶o̶n̶s̶ ̶f̶o̶r̶ ̶e̶a̶c̶h̶ ̶a̶r̶g̶/̶r̶e̶s̶ ̶t̶y̶p̶e̶

Other than that LGTM, spotted a few exported funcs that could use a description for consistency.

EDIT: scratch that first part, making it work would do more harm than good

**Review by katiehockman — APPROVED** [review]
Didn't review the appsec code, but I don't see any major changes in `contrib` other than using a different package, so I'll trust that others have reviewed the rest.

File: internal/appsec/dyngo/operation.go [hunk 7]
```diff
@@ -141,107 +129,106 @@ func NewRootOperation() Operation {
...
 }
 
-// FinishOperation finishes the operation along with its results and emits a
+// Finish finishes the operation along with its results and emits a
 // finish event with the operation results.
 // The operation is then disabled and its event listeners removed.
-func FinishOperation(op Operation, results interface{}) {
-	op.finish(op, results)
-}
+func Finish[O Operation, E ResultOf[O]](op O, results E) {
```
Comment (Julio-Guerra) [inline internal/appsec/dyngo/operation.go:160]: `StartOperation` vs `Finish`: do you want to align both?

**Review by Julio-Guerra — CHANGES_REQUESTED** [review]

**Review by Julio-Guerra — APPROVED** [review]

**Comment by Julio-Guerra** [comment]
Congrats - major API change embracing the Go generics the right way 👏 This was envisioned in the original PR design as a major improvement :-)

**Review by Julio-Guerra — APPROVED** [review]

**Pull request merged**